core/src/main/java/contacts/core/CrudApi.kt

```diff
@@ -0,0 +1,135 @@
+package contacts.core
+
+import android.content.ContentResolver
+import contacts.core.accounts.AccountsPermissions
+import contacts.core.entities.custom.CustomDataRegistry
+
+// ## Developer notes
+//
+// This should actually be an internal interface. It is public because public APIs implement this.
+// Anyways, this enforces some level of uniformity for all CRUD APIs. However, we should try to
+// keep the contract loose to allow for the most flexibility. Who knows? Consumers might find some
+// use for this. There is no howto page for this until we can find an official use for it.
+
+/**
+ * All CRUD (Insert*, Query*, Update*, Delete*) APIs, accessible via a [Contacts], must implement
+ * this interface.
+ */
+interface CrudApi : Redactable {
+
+    /**
+     * A reference to the [Contacts] instance that constructed this. This is mostly used internally
+     * to shorten internal code.
+     *
+     * Don't worry, [Contacts] does not keep references to instances of this. There are no circular
+     * references that could cause leaks =). [Contacts] is just a factory.
+     */
+    val contactsApi: Contacts
+
+    /**
+     * The API core function result.
+     */
+    interface Result : Redactable
+
+    /**
+     * Get notified about events within [CrudApi] instances.
+     */
+    interface Listener {
```

Comment:
This will be very useful for logging; #144 and #145. Actually, I came up with all of this to make it easier to add support for logging ❤️